### PR TITLE
Speculatively implement LWG-3636: `formatter<T>::format` should be const-qualified

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes DevCom-1662387

### CODE REVIEW

**Review by AlexGuteniev — CHANGES_REQUESTED** [review]
When making changes per non-WP-status lwg issue, you must comment every occurence, mention "per LWG-XXX"

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

**Comment by CaseyCarter** [comment]
> When making changes per non-WP-status lwg issue, you must comment every occurence, mention "per LWG-XXX"

Technically, we require commenting behavior that contradicts the Working Draft. All of these `const` additions technically don't contradict the Working Draft; it requires some expressions to be valid, they can be made so regardless of whether `format` is `const`-qualified.

If we were to start calling `format` on `const` objects, that would be a different story.

File: stl/inc/format [hunk 1]
```diff
@@ -2897,7 +2897,7 @@ struct _Formatter_base {
     }
 
 private:
-    _Dynamic_format_specs<_CharT> _Specs;
+    mutable _Dynamic_format_specs<_CharT> _Specs;
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/format:2900]: Technically, as long as no physical modifications are performed, there is no data race. However, I agree that we shouldn't be using `mutable` here, even if that requires more layers of `const` to be added.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:2900]: Technically, even physical modifications would be fine if we used some synchronization mechanism to ensure they would not conflict. In practice, Lines 2884 and 2889 modify this mutable subobject from a `const` member function without any synchronization. I think the fix is to copy `_Specs` (`_Dynamic_format_specs` is small, isn't it?) into a local before any potential modifications in `format`.

If copying is expensive, we might want to do something more complicated like copying only if one of the dynamic indices is non-negative and using a pointer to track if we're using the local specs or object-wide specs.

[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:2900]: +1 on doing the copy, I think that's what libfmt does, and it's probably better than synchronization.

**Review by barcharcraz — CHANGES_REQUESTED** [review]
Looks good, but the mutable is not good very bad.

Unfortunately we need to just copy _Specs into format before getting the dynamic parameters. 

It's not a huge deal as this is only for user-defined types, and if it's a performance problem we can keep the non-const version (assuming wg21 doesn't prohibit that)

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I pushed a stylistic rename (we generally avoid `_UglyCamels` for local variables; changing the pre-existing parameter name here would be a more invasive change and I don't want to unnecessarily interfere with merging) in the hopes of getting the CLA Bot to wake up. Doesn't seem to have helped, though. May need to try again later.

**Review by barcharcraz — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
The CLA bot is also stuck on this PR - @fsb4000 can you close it and open a new PR? We'll get it insta-approved :smile_cat:

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed. (I'll handle verifying that this works with the recent LLVM update.)

**Comment by StephanTLavavej** [comment]
Thanks for time-traveling from the future when this LWG issue has been accepted! :watch: :joy_cat: :white_check_mark:

**Pull request merged**